File: src/main/scala/au/csiro/pbdava/ssparkle/common/utils/LoanUtils.scala
package au.csiro.pbdava.ssparkle.common.utils

import java.io.Closeable
import scala.io.Source

object LoanUtils {

  def withSource[R](src: Source)(func: Source => R):R = {
    try {
      func(src)
    } finally {
      src.close()
    }    
  }
  
  def withCloseable[C <: Closeable,R](cl:C)(func:C => R):R = {
    try {
      func(cl)
    } finally {
      cl.close()
    }
  }  
}

File: src/main/java/au/csiro/sparkle/common/LoanUtils.java
package au.csiro.sparkle.common;

import java.io.Closeable;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;

import org.apache.commons.io.IOUtils;

public final class LoanUtils {

	public static <C extends Closeable> void withCloseable(final C in, ThrowingConsumer<C> func) {
		try {
			func.accept(in);
		} catch (Exception ex) {
			throw new RuntimeException(ex);		
		} finally {
			IOUtils.closeQuietly(in);
		}		
	}

	public static <C extends Closeable, T> T withCloseableFunc(final C in, ThrowingFunction<C,T> func) {
		try {
			return func.apply(in);
		} catch (Exception ex) {
			throw new RuntimeException(ex);
		} finally {
			IOUtils.closeQuietly(in);
		}		
	}
	
	public static void withPrintWriter(final File in, ThrowingConsumer<PrintWriter> func) {
		try {
	        LoanUtils.withCloseable(new PrintWriter(new FileWriter(in)), func);
        } catch (IOException ex) {
	        throw new RuntimeException(ex);
        }		
	}

}
